### PR TITLE
APIGOV-17769 - add check for agentResourceType already a standard generic resource, if not make it one

### PR DESCRIPTION
**Pull request by kflan-io**

### CODE REVIEW

File: pkg/agent/agent.go [hunk 1]
```diff
@@ -319,7 +319,12 @@ func getAgentResource() (*apiV1.ResourceInstance, error) {
 
 	response, err := agent.apicClient.ExecuteAPI(coreapi.GET, agentResourceURL, nil, nil)
 	if err != nil {
```
Comment (sbolosan) [inline pkg/agent/agent.go:321]: this is throwing my off...  why not just return nil, err if err!= nil...
then you can take the if agentResource out of the if and make it stand alone

File: pkg/agent/agent.go [hunk 2]
```diff
@@ -319,7 +319,12 @@ func getAgentResource() (*apiV1.ResourceInstance, error) {
...
-		return nil, err
+		// if agentResourceType is already a standard generic agent resource
+		if agentResourceType == agentTypesMap[agent.cfg.AgentType] {
```
Comment (sbolosan) [inline pkg/agent/agent.go:323]: or move this up under line 317?

File: pkg/agent/agent.go [hunk 3]
```diff
@@ -319,7 +319,12 @@ func getAgentResource() (*apiV1.ResourceInstance, error) {
...
+			return nil, err
+		}
+		AgentResourceType = ""
+		return getAgentResource()
```
Comment (dfeldick) [inline pkg/agent/agent.go:327]: Isn't this calling itself?

File: pkg/agent/agent.go [hunk 4]
```diff
@@ -319,7 +319,12 @@ func getAgentResource() (*apiV1.ResourceInstance, error) {
 
 	response, err := agent.apicClient.ExecuteAPI(coreapi.GET, agentResourceURL, nil, nil)
 	if err != nil {
-		return nil, err
+		// if agentResourceType is already a standard generic agent resource
+		if agentResourceType == agentTypesMap[agent.cfg.AgentType] {
+			return nil, err
+		}
+		AgentResourceType = ""
```
Comment (dfeldick) [inline pkg/agent/agent.go:326]: You maybe didn't do this, but to me code gets confusing if you have variables w/ the same name differing in only the leading upper/lower character. It makes it look like a typo.

File: pkg/agent/agent.go (hunk 1 above)
Comment (jcollins-axway) [reply pkg/agent/agent.go:321]: so this is what Kevin and I worked on together.  AWS and Edge set their resource type to AWSDA/TA, EdgeDA/TA.  If only a generic resource was created an error here is hit that the resource was not found.

We wanted a way to say, lets also check if the agent name give is a generic type before throwing that error.  So we set the type to empty and call this method again, returning the generic resource type call.

File: pkg/agent/agent.go (hunk 2 above)
Comment (jcollins-axway) [reply pkg/agent/agent.go:323]: this prevents an infinite loop, if an error is hit trying to get a generic agent we return the error an ddon't continuously try

File: pkg/agent/agent.go (hunk 3 above)
Comment (jcollins-axway) [reply pkg/agent/agent.go:327]: yes, but resetting the resource type so we look for the generic type rather than dataplane specific type

Comment (dfeldick) [reply pkg/agent/agent.go:327]: Probably deserves a big, fat comment in the code then. It certainly isn't self-documenting as it is.

Comment (jcollins-axway) [reply pkg/agent/agent.go:327]: Agreed

**Review by vivekschauhan — APPROVED** [review]

**Pull request merged**